src/contrib/mlir/lowerer.cpp

```diff
@@ -530,6 +530,102 @@ namespace
+        // We'll keep track of the slice offsets via concatenation_axis_pos.
+        IndexHandle concatenation_axis_pos(index_t(0));
+
+        for (auto& operand : operands)
+        {
+            // Retrieve and check some type info.
+            NGRAPH_CHECK(operand, "Unexpected null operand in ConcatOp");
+            auto operand_ty = operand->getType().dyn_cast<MemRefType>();
+            NGRAPH_CHECK(operand_ty, "Unexpected non-memref operand type");
+            NGRAPH_CHECK(elem_ty == operand_ty.getElementType(), "Types mismatch in ConcatOp");
+
+            // TODO(amprocte): Check consistency of operand shape (rank, dims other than at
+            // concatenation_axis) with result.
+
+            // Assuming rank = r, and the concatenation axis is A where A<r, we'll be creating
+            // loops of this form:
+            //
+            //   for i_0 := 0 to operand_dim[0]:
+            //    for i_1 := 0 to operand_dim[1]:
+            //     ...
+            //      for i_(r-2) := 0 to operand_dim[r-2]:
+            //       for i_(r-1) := 0 to operand_dim[r-1]:
+            //        result[i_0][i_1]...
+            //              [i_(A-1)][i_A + concatenation_axis_pos][i_(A+1)]...
+            //              [i_(r-2)][i_(r-1)]
+            //                  :=
+            //        operand[i_0][i_1]...[i_(r-2)][i_(r-1)]
+            std::vector<IndexHandle> res_index_handles;
+            std::vector<IndexHandle> operand_index_handles;
+            MemRefView v_operand(operand);
+
+            // Function to recursively build the nested loops.
+            size_t axis = 0;
+            std::function<void()> make_copy_loop = [&] {
+                if (axis < rank)
+                {
+                    IndexHandle i;
+                    IndexHandle i_lb(v_operand.lb(axis));
+                    IndexHandle i_ub(v_operand.ub(axis));
+                    int64_t i_step = v_operand.step(axis);
+
+                    LoopBuilder(&i, i_lb, i_ub, i_step)([&] {
+                        res_index_handles.push_back(axis == concatenation_axis
+                                                        ? IndexHandle(i + concatenation_axis_pos)
+                                                        : i);
+                        operand_index_handles.push_back(i);
+                        axis++;
+                        make_copy_loop();
+                    });
+                }
+                else
+                {
+                    IndexedValue iv_operand(operand);
+                    iv_res(res_index_handles) = iv_operand(operand_index_handles);
+                }
+            };
+
+            make_copy_loop();
+
+            // Move up concatenation_axis_pos for the next operand.
+            concatenation_axis_pos = concatenation_axis_pos + v_operand.ub(concatenation_axis);
+        }
+
+        rewriter.replaceOp(op, {result});
+
+        return matchSuccess();
+    }
+
```

Comment:
Would you mind using MLIR's camel naming convention in this file? I still have to fix this in this file and add an note with what he discussed about it.

Reply:
Will do